addons/sms/models/sms_sms.py

```diff
@@ -159,59 +158,49 @@ def _split_batch(self):
-                unlink_failed=unlink_failed, unlink_sent=unlink_sent)
+            results = [{'uuid': sms.uuid, 'state': 'server_error'} for sms in self]
         else:
-            _logger.info('Send batch %s SMS: %s: gave %s', len(self.ids), self.ids, iap_results)
-            self._postprocess_iap_sent_sms(iap_results, unlink_failed=unlink_failed, unlink_sent=unlink_sent)
-
-    def _postprocess_iap_sent_sms(self, iap_results, failure_reason=None, unlink_failed=False, unlink_sent=True):
-        todelete_sms_ids = []
-        if unlink_failed:
-            todelete_sms_ids += [item['res_id'] for item in iap_results if item['state'] != 'success']
-        if unlink_sent:
-            todelete_sms_ids += [item['res_id'] for item in iap_results if item['state'] == 'success']
-
-        for state in self.IAP_TO_SMS_STATE.keys():
-            sms_ids = [item['res_id'] for item in iap_results if item['state'] == state]
-            if sms_ids:
-                if state != 'success' and not unlink_failed:
-                    self.env['sms.sms'].sudo().browse(sms_ids).write({
-                        'state': 'error',
-                        'failure_type': self.IAP_TO_SMS_STATE[state],
-                    })
-                if state == 'success' and not unlink_sent:
-                    self.env['sms.sms'].sudo().browse(sms_ids).write({
-                        'state': 'sent',
-                        'failure_type': False,
-                    })
-                notifications = self.env['mail.notification'].sudo().search([
-                    ('notification_type', '=', 'sms'),
-                    ('sms_id', 'in', sms_ids),
-                    ('notification_status', 'not in', ('sent', 'canceled')),
-                ])
-                if notifications:
-                    notifications.write({
-                        'notification_status': 'sent' if state == 'success' else 'exception',
-                        'failure_type': self.IAP_TO_SMS_STATE[state] if state != 'success' else False,
-                        'failure_reason': failure_reason if failure_reason else False,
-                    })
-        self.mail_message_id._notify_message_notification_update()
-
-        if todelete_sms_ids:
-            self.browse(todelete_sms_ids).sudo().unlink()
+            _logger.info('Send batch %s SMS: %s: gave %s', len(self.ids), self.ids, results)
+        self._postprocess_iap_sent_sms(results, unlink_failed=unlink_failed, unlink_sent=unlink_sent)
+
+    def _postprocess_iap_sent_sms(self, iap_results, unlink_failed=False, unlink_sent=True):
+        mail_message_ids_sudo = self.env['mail.message'].sudo()
+        for iap_state, results in tools.groupby(iap_results, key=lambda result: result['state']):
+            sms_ids_sudo = self.env['sms.sms'].sudo().search([('uuid', 'in', [result['uuid'] for result in results])])
+            mail_message_ids_sudo |= sms_ids_sudo.mail_message_id
```

Comment:
To verify if this will fetch the mail_message_id one by one or if it fetches all at once when calling "_notify_message_notification_update".

In which case it would probably be better to accumulate sms.sms records and at the end of the method do "sms_records.mail_message_id._notify_message_notification_update()".